Feature/allow editable layouts

memsharded opened the pull request:
Changelog: Omit
Docs: Omit

The documentation is in https://github.com/conan-io/docs/pull/1009, but already belongs to the main PR of this feature.

This PR does several things to prepare editables for workspaces:
- Use a unified file ``EDITED_PACKAGES = 'edited_packages'`` to store links, instead of one sentinel per package. This allows workspaces to easily define links at runtime without actually creating hard links
- Allows defining different ``layouts``, with the ``conan link . <ref> --layout=windows``. This is necessary to have editable packages that you could edit in different OSs.
- Layout different files apply both to the cache and to user (edited) repos.
- Allow layout files to be copied and reused between the cache and the repos. That is, the syntax ``[includedirs]`` in the cache is also valid, and applies to all editable packages.
- Do not remove from ``cpp_info`` empty files when they are editable, otherwise, it will fail in packages that haven't been built yet, but consumer packages in same workspace will use.
- Manage the ``cpp_info.xxxxdirs`` as relative whenever possible, not absolute. Transforming them to absolute is a responsibility of the ``cpp_info.xxxx_paths`` methods, that are the ones used by the consumers, generators, etc.
- Raising when an unexpected field is found in layout files like ``[includedrs]`` (which is a typo that can go undetected).
- Raising earlier if parsing a layout file in a repo and it contains namespaces.

SSE4 commented on conans/client/cache.py:
```diff
@@ -319,17 +324,44 @@ def package_summary_hash(self, pref):
+            layout = self._editable_cpp_info[layout_name]
+        except KeyError:
+            layout_file = os.path.join(self.conan_folder, LAYOUTS_FOLDER, layout_name)
+            layout = EditableCppInfo.load(layout_file, True) if os.path.exists(layout_file) else None
```
IMO `os.path.isfile` fits better here (e.g. what if it's a directory?)

SSE4 commented on conans/paths/package_layouts/package_editable_layout.py:
```diff
@@ -25,8 +26,14 @@ def conanfile(self):
 
-    def editable_package_layout_file(self):
-        return os.path.join(self.conan(), CONAN_PACKAGE_LAYOUT_FILE)
+    def layout_file(self):
```
property?

jgsogo commented on conans/test/functional/editable/forbidden_commands_test.py:
```diff
@@ -24,6 +24,7 @@ def setUp(self):
         self.t = TestClient(servers=self.servers, users={"default": [("lasote", "mypass")]})
         self.t.save(files={'conanfile.py': self.conanfile,
                            CONAN_PACKAGE_LAYOUT_FILE: "", })
+        self.t.run('export . {}'.format(self.ref))
```
why?

memsharded replied:
Otherwise, there is nothing in the local cache to be removed, and the test doesn't raise, because it doesn't try to removing anything.

memsharded replied:
Fair, I will decouple the offending test

jgsogo commented on conans/test/unittests/model/editable_cpp_info/apply_test.py:
```diff
@@ -25,63 +25,49 @@
+        self.assertListEqual(cpp_info.libdirs, ['dirs/libdirs'])
+        self.assertListEqual(cpp_info.resdirs, ['dirs/resdirs'])
+        # The default defined by package_info() is maintained
+        self.assertListEqual(cpp_info.bindirs, ['bin'])
```
I think it shouldn't be maintained. The user won't expect his/her layout to be mixed with any other thing.

memsharded replied:
I think it should be respected, specially when users are requesting to be able to define their layout also in the ``package_info()``. If a package_info() defines a layout that is valid for edition wrt the library paths, but then the editable package only needs to map, for example the ``includedirs``, then, you loose all the logic already implemented in ``package_info()``. Lets discuss with @lasote

lasote replied:
Mhh and think I'm ok keeping the cpp_info as a default, but I don't understand why `dirs/bindirs` is not there.

lasote replied:
Oh! now I see. Looks good for me.

memsharded replied:
I understand the reasons.

I changed it because when I was using the feature I really expected the opposite. Why Conan would remove my ``includedirs`` that is already pointing to the ``include`` folder which is always valid for my layout (either cache or editable), and will remove it, and will force me to repeat myself, adding it to the layout file (and to every layout file for every configuration Windows, Linux, etc).

But I agree this is totally a matter of UX, and no problem to discuss, and I am fine with either approach, as long as we agree explicitly on that behavior.

memsharded replied:
Please @uilianries @SSE4 your feedback (as heavy conan users) on what would you expect from this would be very valuable. In short, these are the 2 options:

A) Once a layout file is defined, it will cancel all the ``xxxdirs`` that are defined by ``package_info()``
B) Only what is defined in the layout file overwrites what is defined by ``package_info()``

SSE4 replied:
I also like approach A, as it's more explicit, but for me it appears like we have an inconsistency between `package_info` and `layout` abilities.
for instance, with `package_info`, I may define per-configuration layouts, like:
```
self.cpp_info.defines = ["ALL_CONFIGS"]
self.cpp_info.debug.defines = ["DEBUG"]
self.cpp_info.release.defines = ["NDEBUG"]
```
however, it doesn't appear like I can achieve the same effect with layouts. or am I missing something?
if it's the case, probably users will still want to be able to define per-configuration layouts.

memsharded replied:
Because defines, cppflags, etc, they don't change. Only the "layout" of the folders, need to be re-mapped. Same with library names the library names will be mostly reused from ``package_info()``.

SSE4 replied:
yeah, but what about this:
```
self.cpp_info.libdirs= ["libs_common"]
self.cpp_info.debug.libdirs= ["libs_debug"]
self.cpp_info.release.libdirs= ["libs_release"]
```
can this be defined via layouts?

SSE4 replied:
and the next question then, will definition of `libdirs` in layout invalidate `debug.libdirs` and `release.libdirs` from package info?

SSE4 replied:
yeah, that's my point - users will probably expect feature parity between `layout` and `package_info`, otherwise, layout will override/invalidate things that it cannot handle without providing an easy way to avoid it

memsharded replied:
That is the point of my proposal:
- It is very likely that in most cases, the normal compilation information defined in ``package_info()``, like definitions, flags, library names, etc. won't be, and probably shouldn't be re-defined by a layout file
- The layout files only affect by definition to paths: ``includedirs`` and family
- Then, the final result will always be a mixture of 2 sources of information: the ``package_info()`` method, and the ``layout`` file.
- If we don't do this, there will be immediate complains about users that need to repeat themselves.
- Note that the layout files do not contain any logic, it is necessary to switch and use a different file if you want to change the layout. It is typical that some definitions are conditional, like those BOOST_USE_DYNAMIC-like when using shared libraries, so it would be a nightmare to have to define those in layout files.
- Adding logic to layouts was done in conan workspaces, with ``eval()``. While doable, it is a new syntax that users need to learn. Maybe we want to go this way instead of switching between layout files?

SSE4 replied:
I am not sure, to be honest. e.g. how does switching layout matches with typical developer's workflow, e.g. someone who uses IDE like Visual Studio, CLion, XCode, etc. and tries to switch between build configurations in his favorite IDE.
now developer also has to remember that he has to switch layout each time he switches configuration. or probably, somehow to do it automatically (but how?).
okay, now let's switch to the command line use-case, where developer just uses bunch of conan commands, or may be even conan package tools, how now he should always keep in mind to switch layouts in time? should we add arguments to the existing commands to accept layout, so they call `conan link` internally (just like in case of `--profile` argument)? should we make CPT aware about this new feature?

jgsogo replied:
IMO there are two different things:
 
* There are _placeholders_ available in the layout files, you can write something like `mydir/{settings.build_type}/folder` and it will be resolved to `mydir/Debug/folder` or `mydir/Release/folder` (same with options and the other settings). I'm sure this is not a solution, and won't fit user needs, but it declares a path to walk. If we need conditional logic inside the layout files I will go for a template system instead of python `eval`, I find it more powerful and maintainable.

 * A completely different thing is being able to specify a file when linking a package as editable. I find it quite useful, but IMO it is not for linking debug-release (previous point) or linux-win, but to use a file for a library, a different file for another library,... and let the user define those files. Up to now we can just use the file inside the repo or the `layouts/default` one, which is not fair.

memsharded replied:
I have added a new functional test to validate the placeholders approach might be valid for the Debug/Release case.

The idea is that the different layout files are to differentiate between different OS or build systems, while the placeholders should be good to handle different configurations within the same OS and build system, i.e. the different Debug/Release, x86/64, etc. typical variants.

jgsogo commented on conans/test/unittests/model/editable_cpp_info/parse_test.py:
```diff
@@ -5,96 +5,41 @@
+                            something
+                            """)
+        save(self.layout_filepath, content)
+        with self.assertRaisesRegexp(ConanException, "Wrong cpp_info field: includedrs"):
```
Is this exception catch elsewhere and more context added to it? Things like: this is about a layout for editable files, this file is located in this path,...

memsharded replied:
I am not sure I understand your comment. If the file that defines the layout contains unexpected syntax, shouldn't it raise?

jgsogo commented on conans/test/unittests/model/editable_cpp_info/parse_test.py:
```diff
@@ -5,96 +5,41 @@
+                            something
+                            """)
+        save(self.layout_filepath, content)
+        with self.assertRaisesRegexp(ConanException, "Repository layout file doesn't allow pattern"):
```
This is the reader, it knows nothing about where the file is located (inside/outside the repo). This message is assuming some implicit information.

memsharded replied:
Oh, I see. Yes, lets add the path to the offending file.

jgsogo commented on conans/test/unittests/model/editable_cpp_info/work_on_items_test.py:
```diff
@@ -1,80 +1,32 @@
+        self.assertEqual('/usr/path with spaces/Visual Studio/dir',
+                         EditableCppInfo._work_on_item("/usr/path with spaces/{settings.compiler}/dir",
+                                                       settings=settings,
+                                                       options=None))
```
More tests:
 * windows-like with just one backward slash
 * path with an ending slash -> I think that this ending slash should be removed (or added one to every other path)

jgsogo commented on conans/client/command.py:
```diff
@@ -1342,6 +1342,8 @@ def link(self, *args):
         parser.add_argument("--remove", action='store_true', default=False,
                             help='Remove linked reference (target not required)')
+        parser.add_argument("-l", "--layout",
+                            help='Define an alternative layout filename')
```
IMO, this should be a path to a file, not just a name. Maybe we can agree that relative paths are resolved to the _layouts_ folder inside the cache (here you have the _layout name_), but this should be converted to an absolute path as soon as possible and this path stored associated with the reference. And by default, this argument should take the _default layout_ name/path.

The logic in the installer will be straightforward then.

memsharded replied:
Tested, this arguments allows relative paths to files, both in the edited repo and in the local cache.

SSE4 replied:
I think we already had misleading situations when profile name incorrectly resolved to the file with the same name incidentally appeared in the same folder during build, which was hard to debug

memsharded replied:
Shouldn't that case just raise an error saying that the file is incorrect?
I could add tests with random other files to check error messages

SSE4 replied:
probably, better to fail early if such coincidence happen, at least it will help to debug

memsharded replied:
Added a test for outputing a clear error when another file is parsed.

jgsogo replied:
Need to check what it is stored in the cache, I would want to see there an absolute path (easier for everything I think).

memsharded commented on conans/client/cache.py:
```diff
@@ -273,17 +279,44 @@ def invalidate(self):
-        linked_folder_sentinel = self._build_path_to_linked_folder_sentinel(ref)
-        save(linked_folder_sentinel, content=target_path)
+    def _load_edited(self):
+        edited_path = normpath(join(self.conan_folder, EDITED_PACKAGES))
```
Totally agree, will do it.

Will leave moving the files to a dedicated python package for a different PR, to ease review.

jgsogo replied:
I think that this comment still stands, better `editing` or `editable`, but not `edited`.

memsharded commented on conans/test/functional/editable/layouts_test.py:
```diff
@@ -0,0 +1,83 @@
+            """)
+        client2.save({"conanfile.txt": consumer})
+        client2.run("install .", assert_error=True)
+        self.assertIn("ERROR: Layout file not found: missing", client2.out)
```
What if exists at link time, but then it is removed when you actually need it? Or you can link and add it later the layout file. Also, it might find the repo file, but then you remove it, and then it will find a cache layout one. This check is totally necessary, irrespective if other check could be introduced at link time, but I think that one gives almost no value.

jgsogo replied:
I agree with @danimtb, I think that the check is very useful to avoid errors (misspellings, confusion resolving relative paths,...) at the moment of writing the `conan link` command. This is aligned with my comment about the `--layout` argument: https://github.com/conan-io/conan/pull/4287#discussion_r247824870

memsharded replied:
Added this check now ``conan link --layout`` will raise if the layout file is not found or incorrect.

lasote commented:
What about the docs, something needed?

memsharded commented:
Bunch of changes, please review again:

- Not using a local default layout at all. If some local default file is desired, needs to be explicit in: ``conan link --layout=layoutfile``
- Changed paths, they are relative to the current directory, not the recipe directory
- Once a match is found in a layout file, then the xxx_dirs folders defined in ``package_info()`` are emptied.
- Added success message printing the used layout file
- Added tests for setting local edition layout directly in ``package_info()`` (check ``in_local_cache``)
- Allow full references in the ``[Pkg/0.1@user/testing:includedirs]``
- Allowing namespace/scope in all contexts of layout files, as even repo files can be reused in other packages.
- Removed wildcard * in layout files

jgsogo commented on conans/client/command.py:
```diff
@@ -1342,6 +1342,8 @@ def link(self, *args):
         parser.add_argument("--remove", action='store_true', default=False,
                             help='Remove linked reference (target not required)')
+        parser.add_argument("-l", "--layout",
+                            help='Define an alternative layout filename')
```
Improve help message

jgsogo commented on conans/client/edited.py:
```diff
@@ -0,0 +1,44 @@
+from conans.util.files import load, save
+
+
+EDITED_PACKAGES = 'edited_packages'
```
If this is a file, name according to it:

```suggestion
EDITED_PACKAGES_FILE = 'edited_packages'
```

jgsogo commented on conans/model/editable_cpp_info.py:
```diff
@@ -1,73 +1,73 @@
+        try:
+            parser.read(filepath)
+        except configparser.Error:
+            raise ConanException("Error parsing layout file: %s" % filepath)
```
May add the error message from the exception if it adds more information to the user, what kind of messages do we get in the `configparser.Error` exception?

memsharded replied:
Yes, something might be useful. Adding exception.

jgsogo commented on conans/model/editable_cpp_info.py:
```diff
@@ -1,73 +1,73 @@
+            if key not in EditableCppInfo.cpp_info_dirs:
+                raise ConanException("Wrong cpp_info field '%s' in layout file: %s"
+                                     % (key, filepath))
+            data.setdefault(pkg, {})[key] = [k for k, _ in parser.items(section)]
```
We are calling this function to test the file format, does it make sense to test the contents? We should check that every item is a valid folder after substituting the placeholders (maybe in a different function).

jgsogo commented on conans/model/editable_cpp_info.py:
```diff
@@ -1,73 +1,73 @@
+            raise ConanException("Error parsing layout file: %s" % filepath)
+        data = OrderedDict()
+        for section in parser.sections():
+            pkg, key = section.split(":", 1) if ':' in section else (None, section)
```
rename: `pkg` >> `reference`

jgsogo commented on conans/model/editable_cpp_info.py:
```diff
@@ -1,73 +1,73 @@
-            data_to_apply = self._data
+    def apply_to(self, ref, cpp_info, settings=None, options=None):
+        d = self._data
+        data = d.get(str(ref)) or d.get(None) or {}
```
this is called using `ref=node.ref`, does this `str(ref)` include the revision?

memsharded replied:
No, the str (repr) doesn't use the revision at all.

jgsogo commented on conans/paths/package_layouts/package_editable_layout.py:
```diff
@@ -25,25 +27,28 @@ def conanfile(self):
-        raise RuntimeError("Operation not allowed on a package installed as editable")
 
     def package_metadata(self):
         # FIXME: I know that downstream there is an except for IOError
```
Please, review the exception that is being thrown from this function, we have to take a look at the places where the `package_metadata` is being called

memsharded replied:
Added exception again

jgsogo replied:
The exception is there, but it is an `IOError` exception. I really dislike throwing from this function an `IOError` exception (when everything else is a `ConanException`) just because the context this function is called:

```python
    # Metadata
    def load_metadata(self):
        try:
            text = load(self.package_metadata())
            return PackageMetadata.loads(text)
        except IOError:
            return PackageMetadata()
```

I'm ok with it if we are ok with it. I was just asking for confirmation or maybe modify that function to catch a ConanException too and throw it from here.

jgsogo commented on conans/paths/simple_paths.py:
```diff
@@ -40,25 +39,11 @@ def __init__(self, store_folder):
-        linked_package_file = os.path.join(base_folder, LINKED_PACKAGE_SENTINEL)
-        return linked_package_file
-
     def package_layout(self, ref, short_paths=False):
```
This function is overriden in the `ClientCache` class, I think that this implementation is no longer used.

memsharded replied:
It is being used by the server, that is not using ClientCache, but SimplePaths class

jgsogo commented on conans/test/functional/editable/link_create_test.py:
```diff
@@ -22,10 +21,17 @@ class APck(ConanFile):
         ref = ConanFileReference.loads('lib/version@user/name')
         t = TestClient()
-        t.save(files={'conanfile.py': self.conanfile, CONAN_PACKAGE_LAYOUT_FILE: ""})
+        t.save(files={'conanfile.py': self.conanfile, "mylayout": ""})
```
```suggestion
        t.save(files={'conanfile.py': self.conanfile})
```

jgsogo commented on conans/test/functional/editable/link_remove_test.py:
```diff
@@ -21,8 +20,8 @@ def setUp(self):
-                           CONAN_PACKAGE_LAYOUT_FILE: "", })
-        self.t.run('link . {}'.format(self.ref))
+                           "mylayout": "", })
+        self.t.run('link . {} -l=mylayout'.format(self.ref))
```
Check the output: `linked to path ....`

memsharded replied:
Not the correct place, a setUp() for a remove test suite.

The "linked to path" functionality is tested in a "link" test.

jgsogo replied:
Totally agree!

jgsogo commented on conans/test/unittests/model/editable_cpp_info/parse_test.py:
```diff
@@ -5,96 +5,32 @@
-        base_content.format(namespace="libA:", path_prefix="libA/")
-    ])
 
+class ParseTest(unittest.TestCase):
```
We need here:
 * parsing a pattern -> fail
 * parsing a full reference with revision -> fail

jgsogo commented on conans/client/command.py:
```diff
@@ -1342,6 +1341,10 @@ def link(self, *args):
+        parser.add_argument("-l", "--layout",
+                            help='Relative or absolute path to a file containing the layout.'
+                            ' Relative paths will be resolved first relative to current dir, '
+                            'then to local cache')
```
to "local cache" or to "<cache>/layouts"?

jgsogo commented on conans/client/edited.py:
```diff
@@ -0,0 +1,44 @@
+import json
```
Change filename to match the renaming from `edited` to `editable`.

memsharded commented:
Not used, I have removed it, and all green. Please @jgsogo review.

memsharded commented:
Docs available in https://github.com/conan-io/docs/pull/1009
